pkg/cmd/container/list.go

```diff
@@ -112,58 +104,101 @@ func (x *ListItem) Label(s string) string {
 	return x.LabelsMap[s]
 }
 
-func prepareContainers(ctx context.Context, client *containerd.Client, containers []containerd.Container, options types.ContainerListOptions) ([]ListItem, error) {
+func prepareContainers(ctx context.Context, client *containerd.Client, containers []containerd.Container, options types.ContainerListOptions, all bool) ([]ListItem, error) {
 	listItems := make([]ListItem, len(containers))
 	snapshottersCache := map[string]snapshots.Snapshotter{}
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(containers))
 	for i, c := range containers {
-		info, err := c.Info(ctx, containerd.WithoutRefreshedMetadata)
-		if err != nil {
-			if errdefs.IsNotFound(err) {
-				log.G(ctx).Warn(err)
-				continue
+		wg.Add(1)
+		go func(ctx context.Context, c containerd.Container, i int) {
+			defer wg.Done()
+			info, err := c.Info(ctx, containerd.WithoutRefreshedMetadata)
+			if err != nil {
+				if errdefs.IsNotFound(err) {
+					log.G(ctx).Warn(err)
+					return
+				}
+				errChan <- err
 			}
-			return nil, err
-		}
-		spec, err := c.Spec(ctx)
-		if err != nil {
-			if errdefs.IsNotFound(err) {
-				log.G(ctx).Warn(err)
-				continue
+			spec, err := c.Spec(ctx)
+			if err != nil {
+				if errdefs.IsNotFound(err) {
+					log.G(ctx).Warn(err)
+					return
+				}
+				errChan <- err
+			}
+			id := c.ID()
+			if options.Truncate && len(id) > 12 {
+				id = id[:12]
 			}
-			return nil, err
+			cStatus := formatter.ContainerStatus(ctx, c)
+			// only show Up status container
+			if !all {
+				if !strings.HasPrefix(cStatus, "Up") {
+					return
+				}
+			}
```

Comment:
=>

```golang
if !all &&!strings.HasPrefix(cStatus, "Up") {
  return
}
```

Reply:
also, what's the improvement if we mvoe this logic into `filterContainers`?

Now the filter is done in 2 separate functions, and `filterContainers` has to return a `all` to indiciate if it's a `all` filter or not, which is a little confusing.

Reply:
old version the bottleneck is  formatter.ContainerStatus, and call it twice.So delete one is necessary.

Reply:
or use goroutine just in filterContainers func.